Fix MeshImporter to handle attributes that are faceVarying correctly

lucillecaillaud opened the pull request:
* If any attributes is faceVarying: "unroll" the vertices to be the same size of the indices.
* Attributes that are faceVarying doesn't need to be unrolled anymore
* Attributes that are uniform are handled like if they were faceVarying
* MeshSampleBase now have Primvar to store uv values (it should actually be the case also for tangents and colors)

lucillecaillaud committed:
WIP: Fix MeshImporter to handle FaceVarying values correctly

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -603,55 +676,120 @@ public class GeometrySubsets {
-                             int[] faceVertexCounts,
-                             int[] faceVertexIndices) {
-      if (primvars == null || primvars.Count == 0) { return; }
+    private static Vector3[] SanitizePointsForUnity(Vector3[] vertices, int[] indices)
```
Name game: this is not really "sanitizing" points, it's really expanding the indices (or unrolling them?).

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -266,67 +268,84 @@ public class GeometrySubsets {
+        {
+          // TODO: I think that vertex indices and counts are mandatory attributes for a prim with a Mesh schema,
+          //       we could have a generic validation phase for that (generic = uniform checks of mandatory
+          //       attributes for any schema). 
```
Broken assets often are missing data that should be required :/

lucillecaillaud replied:
Yes, but what I mean is that maybe... we could have a function that given a prim and it's schema, validate that all mandatory attributes are valid, and if not, ignore the prim... or something like that... and avoid too much "if blabla != null" in the code.

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -337,64 +356,78 @@ public class GeometrySubsets {
           int curOffset = 0;
-          for (int i = 0; i < usdMesh.faceVertexCounts.Length; i++) {
+          for (int i = 0; i < usdMesh.faceVertexCounts.Length; i++)
+          {
```
Curly braces on new line is actually the Unity style, but I would make that update in a separate change (across the entire code base), since it makes it harder to see the diff in this PR

lucillecaillaud replied:
I know, I'm sorry, I did it for readability when I worked on it, but I should have removed it before doing the PR... I was lazy... I'll do it in my next commit :)

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -266,67 +268,84 @@ public class GeometrySubsets {
+        
+        if (IsThereFaceVaryingAttributes(usdMesh, options))
+        {
+          newVertices = SanitizePointsForUnity(newVertices ?? usdMesh.points, usdMesh.faceVertexIndices);
```
I would add a comment here explaining what's happening and how it affects the rest of the import process below.

jcowles replied:
Also, the faces only need to be split on seams, thought that is much more of a pain to implement.

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -337,64 +356,78 @@ public class GeometrySubsets {
               newFaceIndices.Add(faceIndex + offsetMapping[faceIndex]);
-              for (int i = 1; i < usdMesh.faceVertexCounts[faceIndex] - 2; i++) {
+              for (int i = 1; i < usdMesh.faceVertexCounts[faceIndex] - 2; i++)
+              {
```
You should definitely test an asset that has geometry subsets.

lucillecaillaud replied:
Do we have a known one?

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -422,116 +455,143 @@ public class GeometrySubsets {
+        }
+        else
+        {
+          Color[] sanitizedColors;
```
Name game again: "sanitized" is a little misleading

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -603,55 +676,120 @@ public class GeometrySubsets {
+      int pointCount = usdMesh.points.Length;
+      int faceCount = usdMesh.faceVertexCounts.Length;
+
+      // Potential face varying attributes are: normals, tangents, color and uv (st, uv, uv2. uv3. uv4).
```
Any primvar requested by UsdPreviewSurface also needs to be included....

lucillecaillaud replied:
The `|| options.ShouldBindMaterials` is here for that... It is a dirty shortcut... but I'm not sure how to check all the primvar requested by UsdPreviewSurface?

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -603,55 +676,120 @@ public class GeometrySubsets {
+        {
+          continue;
+        }
         VtVec2fArray vec2fArray = UsdCs.VtValueToVtVec2fArray(val);
```
Oof, I just noticed that we are assuming the primvar is a vec2 here...

lucillecaillaud replied:
Should we change that? How?

jcowles commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Geometry/MeshImporter.cs:
```diff
@@ -740,36 +909,41 @@ public class GeometrySubsets {
-      if (uvVec.Length > vertexCount) {
-        uvVec = UnrollFaceVarying(vertexCount, uvVec, faceVertexCounts, faceVertexIndices);
+      if (uvVec.Length != vertexCount) {
+//        uvVec = UnrollFaceVarying(vertexCount, uvVec, faceVertexCounts, faceVertexIndices);
```
dead comment?

lucillecaillaud replied:
Actually I'm missing some refactoring here as well... I did nothing except commenting the call to the function ^^

jcowles reviewed (commented):
Incredible!!
This looks correct to me, though I'm not 100% sure about geometry subsets.

lucillecaillaud commented on src/USD.NET.Unity/Geometry/MeshSampleBase.cs:
```diff
@@ -37,19 +37,10 @@ public class MeshSampleBase : PointBasedSample {
+    public Primvar<object> uv = new Primvar<object>();
+    public Primvar<object> uv2 = new Primvar<object>();
+    public Primvar<object> uv3 = new Primvar<object>();
+    public Primvar<object> uv4 = new Primvar<object>();
```
Arf... true... Is there a way to keep the backward compatibility?

Or, simply release a minor (or, actually a major) and disclaim that it's not backward compatible with existing script that use that? And, don't group the release of this feature with the one already merged.

lucillecaillaud committed:
PR: Address some comments, improve the "ImportUV" function

lucillecaillaud committed:
PR: Revert changes on bracket formating.

The changes are reverted mostly on line that wasn't changed otherwise
(i.e. remove non-meaningful diff).

lucillecaillaud committed:
MeshImporter: Handle vertex varying primvars in LoadPrimvars

lucillecaillaud committed:
WIP: SkeletonImporter: Unroll skinning information when vertices was unrolled

This is to ensure bones and skinning will match the vertices even when
they were unrolled in the MeshImporter.

To do: Animation in a timeline is still broken, there is still something
to fix there.

To think: This whole "unroll" process is very tedious, we may need to
re-think the process of creating a (skinned) mesh in Unity from the Usd
data.

judubu commented on package/com.unity.formats.usd/Runtime/Scripts/IO/Skel/SkeletonImporter.cs:
```diff
@@ -274,13 +275,35 @@ public static class SkeletonImporter {
+        float[] newWeights = new float[faceVertexIndices.Length * weightsElementSize];
+        for (int i = 0; i < faceVertexIndices.Length; i++)
+        {
+          for (int idxSize = 0; idxSize < indicesElementSize; idxSize++)
```
You can merge both loops together as indicesElementSize always equals weightsElementSize

lucillecaillaud committed:
TypeBinding: Add support for TexCoord2f attributes array.

lucillecaillaud committed:
SceneOptions: keep info of whether attributes value need to be unrolled or not

judubu commented:
The importer is being rewritten.
